atomicapp/providers/lib/kubernetes.py

```diff
@@ -0,0 +1,170 @@
+            raise KubeKubernetesError("Kubernetes API URL does not include HTTP or HTTPS")
+
+        # Gather what end-points we will be using
+        self.k8s_api = urljoin(url, "api/v1/")
+
+        # Test the connection before proceeding
+        self.api.test_connection(self.k8s_api)
+
+        # Gather the resource names which will be used for the 'kind' API calls
+        self.k8s_api_resources = self.api.get_resources(self.k8s_api)
+
+    def create(self, obj, namespace):
+        '''
+        Create an object from the Kubernetes cluster
+        '''
+        name = self._get_metadata_name(obj)
+        kind, url = self._generate_kurl(obj, namespace)
+        self.api.request("post", url, data=obj)
+        logger.info("%s '%s' successfully created" % (kind.capitalize(), name))
+
+    def delete(self, obj, namespace):
+        '''
+        Delete an object from the Kubernetes cluster
+
+        Args:
+            obj (object): Object of the artifact being modified
+            namesapce (str): Namespace of the kubernetes cluster to be used
+            replicates (int): Default 0, size of the amount of replicas to scale
+
+        *Note*
+        Replication controllers must scale to 0 in order to delete pods.
+        Kubernetes 1.3 will implement server-side cascading deletion, but
+        until then, it's mandatory to scale to 0
+        https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/garbage-collection.md
+
+        '''
+        name = self._get_metadata_name(obj)
+        kind, url = self._generate_kurl(obj, namespace, name)
+
+        if kind in ['rcs', 'replicationcontrollers']:
+            self.scale(obj, namespace)
+
+        self.api.request("delete", url)
+        logger.info("%s '%s' successfully deleted" % (kind.capitalize(), name))
+
+    def scale(self, obj, namespace, replicas=0):
+        '''
+        By default we scale back down to 0. This function takes an object and scales said
+        object down to a specified value on the Kubernetes cluster
+
+        Args:
+            obj (object): Object of the artifact being modified
+            namesapce (str): Namespace of the kubernetes cluster to be used
+            replicates (int): Default 0, size of the amount of replicas to scale
+        '''
+        patch = [{"op": "replace",
+                  "path": "/spec/replicas",
+                  "value": replicas}]
+        name = self._get_metadata_name(obj)
+        _, url = self._generate_kurl(obj, namespace, name)
+        self.api.request("patch", url, data=patch)
+        logger.info("'%s' successfully scaled to %s" % (name, replicas))
+
+    def namespaces(self):
+        '''
+        Gathers a list of namespaces on the Kubernetes cluster
+        '''
+        url = urljoin(self.k8s_api, "namespaces")
+        ns = self.api.request("get", url)
+        return ns['items']
+
+    def _generate_kurl(self, obj, namespace, name=None, params=None):
```

Comment:
Yup. I added that to the comments. We could do "gen_kind_url" instead, would this be better?